File: JsonDatabase.py
import os
import json
import logging
import pip
#import pandas as pd

logging.basicConfig(level=logging.INFO)


class DatabaseError(Exception):
    pass

class JsonDatabase:
    def __init__(self, location=".", name="default"):
        self.location = location
        self.name = name

    def create_database(self):
        """Create a new JSON database file if it doesn't exist."""
        database_folder = os.path.join(self.location, "database")
        if not os.path.exists(database_folder):
            os.makedirs(database_folder)
        path = os.path.join(database_folder, self.name + '.json')
        if not os.path.exists(path):
            try:
                with open(path, 'w') as file:
                    file.write("{\n\n}")
                logging.info(f"Created database '{self.name}' at '{path}'.")
            except IOError as e:
                raise DatabaseError(f"Error creating database '{self.name}': {e}")
        else:
            logging.warning(f"Database '{self.name}' already exists at '{path}'.")

    def read_raw_database(self):
        """Read raw content from a JSON database file."""
        database_folder = os.path.join(self.location, "database")
        path = os.path.join(database_folder, self.name + '.json')
        try:
            with open(path, "r") as file:
                content = file.read()
            return content
        except IOError as e:
            raise DatabaseError(f"Error reading database '{self.name}': {e}")

    def read_database(self):
        """Read JSON content from a database file."""
        content = self.read_raw_database()
        try:
            JSONContent = json.loads(content)
            return JSONContent # returns Python dict
        except json.JSONDecodeError as e:
            raise DatabaseError(f"Error decoding JSON in database '{self.name}': {e}")
    
    def insert_into(self, key, data):
        """Insert data into a JSON database."""
        database_folder = os.path.join(self.location, "database")
        existing_data = self.read_database()
    
        # Check if the key already exists in the database
        if key in existing_data:
            logging.warning(f"Key '{key}' already exists in database '{self.name}'. Overwriting existing data.")
    
        # Update the database with the new data
        existing_data[key] = data
    
        # Write the updated data to the database file
        self.write_json(existing_data)
        logging.info(f"Inserted data into '{key}' in database '{self.name}'.")

    def delete_database(self):
        """Delete the JSON database file."""
        database_folder = os.path.join(self.location, "database")
        path = os.path.join(database_folder, self.name + '.json')
        try:
            os.remove(path)
            logging.info(f"Deleted database '{self.name}'.")
        except FileNotFoundError:
            logging.warning(f"Database '{self.name}' does not exist.")
        except IOError as e:
            raise DatabaseError(f"Error deleting database '{self.name}': {e}")

    def delete_key(self, key):
        """Delete a key from a JSON database."""
        database_folder = os.path.join(self.location, "database")
        existing_data = self.read_database()
        if key not in existing_data:
            raise KeyError(f"Key '{key}' not found in the database '{self.name}'.")
        existing_data.pop(key)
        # Write the updated data to the database file
        self.write_json(existing_data)
        logging.info(f"Deleted key '{key}' from database '{self.name}'.")

    def update(self, key, new_data):
        """Update data associated with a key in a JSON database."""
        database_folder = os.path.join(self.location, "database")
        existing_data = self.read_database()
    
        # Check if the key exists in the database
        if key not in existing_data:
            raise KeyError(f"Key '{key}' not found in the database '{self.name}'.")
    
        # Update the existing data with new_data
        if isinstance(existing_data[key], dict) and isinstance(new_data, dict):
            existing_data[key].update(new_data)
        else:
            existing_data[key] = new_data
    
        # Write the updated data to the database file
        self.write_json(existing_data)
        logging.info(f"Updated data for key '{key}' in database '{self.name}'.")

    def write_json(self, python_dict):
        """Write data to a JSON database file."""
        database_folder = os.path.join(self.location, "database")
        path = os.path.join(database_folder, self.name + '.json')
        try:
            with open(path, "w") as file:
                json.dump(python_dict, file)
            #logging.info(f"Updated database '{self.name}'")
        except IOError as e:
            raise DatabaseError(f"Error writing at database '{self.name}': {e}")

    def show(self,toshowORtoreturn="show"):
        """Show database content in tabular form."""
        try:
            import pandas as pd  
        except ImportError:
            pd = None
            logging.warning("Pandas is not installed. For running this command pandas is required. Please install it by 'pip install pandas' or 'py -m pip install pandas'")
            inp = input("Write y/n if want to install/not install pandas")
            if (inp == "y"):
                pip.main(['install', 'pandas'])
            elif (inp == "n"):
                pass
            else:
                logging.error("No valid option selected")
        existing_data = self.read_database()
        df = pd.DataFrame.from_dict(existing_data, orient='index')
        if (toshowORtoreturn == "show"):
            print(df)
        elif(toshowORtoreturn == "return"):
            return df

File: JsonDatabase.js
const fs = require('fs');
const path = require('path');
const jsonfile = require('jsonfile');
const readline = require('readline');

class DatabaseError extends Error {}

class JsonDatabase {
    constructor(location = ".", name = "default") {
        this.location = location;
        this.name = name;
    }

    create_database() {
        const databaseFolder = path.join(this.location, "database");
        try {
            // Ensure the entire directory path exists
            fs.mkdirSync(databaseFolder, { recursive: true });
            console.info(`Created directory at '${databaseFolder}'`);
        } catch (e) {
            throw new DatabaseError(`Error creating directory '${databaseFolder}': ${e}`);
        }

        const filePath = path.join(databaseFolder, this.name + ".json");
        if (!fs.existsSync(filePath)) {
            try {
                fs.writeFileSync(filePath, "{\n\n}");
                console.info(`Created database '${this.name}' at '${filePath}'.`);
            } catch (e) {
                throw new DatabaseError(`Error creating database '${this.name}': ${e}`);
            }
        } else {
            console.warn(`Database '${this.name}' already exists at '${filePath}'.`);
        }
    }

    read_raw_database() {
        const databaseFolder = path.join(this.location, "database");
        const filePath = path.join(databaseFolder, this.name + ".json");
        try {
            return fs.readFileSync(filePath, "utf-8");
        } catch (e) {
            throw new DatabaseError(`Error reading database '${this.name}': ${e}`);
        }
    }

    read_database() {
        const content = this.read_raw_database();
        try {
            return JSON.parse(content);
        } catch (e) {
            throw new DatabaseError(`Error decoding JSON in database '${this.name}': ${e}`);
        }
    }

    insert_into(key, data) {
        const existingData = this.read_database();

        if (key in existingData) {
            console.warn(`Key '${key}' already exists in database '${this.name}'. Overwriting existing data.`);
        }

        existingData[key] = data;
        this.write_json(existingData);
        console.info(`Inserted data into '${key}' in database '${this.name}'.`);
    }

    delete_database() {
        const databaseFolder = path.join(this.location, "database");
        const filePath = path.join(databaseFolder, this.name + ".json");
        try {
            fs.unlinkSync(filePath);
            console.info(`Deleted database '${this.name}'.`);
        } catch (e) {
            if (e.code === 'ENOENT') {
                console.warn(`Database '${this.name}' does not exist.`);
            } else {
                throw new DatabaseError(`Error deleting database '${this.name}': ${e}`);
            }
        }
    }

    delete_key(key) {
        const existingData = this.read_database();
        if (!(key in existingData)) {
            throw new Error(`Key '${key}' not found in the database '${this.name}'.`);
        }
        delete existingData[key];
        this.write_json(existingData);
        console.info(`Deleted key '${key}' from database '${this.name}'.`);
    }

    update(key, newData) {
        const existingData = this.read_database();

        if (!(key in existingData)) {
            throw new Error(`Key '${key}' not found in the database '${this.name}'.`);
        }

        if (typeof existingData[key] === 'object' && typeof newData === 'object') {
            existingData[key] = { ...existingData[key], ...newData };
        } else {
            existingData[key] = newData;
        }

        this.write_json(existingData);
        console.info(`Updated data for key '${key}' in database '${this.name}'.`);
    }

    write_json(data) {
        const databaseFolder = path.join(this.location, "database");
        const filePath = path.join(databaseFolder, this.name + ".json");
        try {
            jsonfile.writeFileSync(filePath, data, { spaces: 2 });
        } catch (e) {
            throw new DatabaseError(`Error writing at database '${this.name}': ${e}`);
        }
    }

    async show(toshowORtoreturn = "show") {
        let pd;
        try {
            pd = require('pandas-js');  // Note: pandas-js is a placeholder for actual tabular data handling in JS
        } catch (e) {
            console.warn("Pandas is not installed. For running this command pandas is required. Please install it by 'npm install pandas-js'");
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout
            });
            const answer = await new Promise(resolve => rl.question("Write y/n if want to install/not install pandas: ", resolve));
            rl.close();
            if (answer.toLowerCase() === 'y') {
                require('child_process').execSync('npm install pandas-js');
                pd = require('pandas-js');
            } else if (answer.toLowerCase() === 'n') {
                return;
            } else {
                console.error("No valid option selected");
                return;
            }
        }

        const existing_data = this.read_database();
        const df = pd.DataFrame(existing_data);
        if (toshowORtoreturn === "show") {
            console.log(df);
        } else if (toshowORtoreturn === "return") {
            return df;
        }
    }
}

module.exports = JsonDatabase;